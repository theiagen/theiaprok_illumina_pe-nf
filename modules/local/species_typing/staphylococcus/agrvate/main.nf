process AGRVATE {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/biocontainers/agrvate:1.0.2--hdfd78af_0"

    input:
    tuple val(meta), path(assembly)

    output:
    tuple val(meta), path("*.agrvate.tsv")    , emit: agrvate_summary
    tuple val(meta), path("*.agrvate.tar.gz") , emit: agrvate_results
    tuple val(meta), path("AGR_GROUP")        , emit: agrvate_agr_group
    tuple val(meta), path("AGR_MATCH_SCORE")  , emit: agrvate_agr_match_score
    tuple val(meta), path("AGR_CANONICAL")    , emit: agrvate_agr_canonical
    tuple val(meta), path("AGR_MULTIPLE")     , emit: agrvate_agr_multiple
    tuple val(meta), path("AGR_NUM_FRAMESHIFTS"), emit: agrvate_agr_num_frameshifts
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def typing_only = params.agrvate_typing_only ? '--typing-only' : ''
    
    """
    # Run agrvate on assembly
    # Using -m flag for mummer frameshift detection since usearch is not available
    agrvate \\
        ${typing_only} \\
        -i ${assembly} \\
        -m \\
        ${args}
    
    # agrvate names output directory and file based on name of .fasta file
    # so <prefix>.fasta as input results in <prefix>-results/ outdir
    # and results in <prefix>-results/<prefix>-summary.tab files
    basename=\$(basename ${assembly})
    # Strip off anything after the period
    fasta_prefix=\${basename%.*}
    
    # Rename outputs summary TSV to include samplename
    mv -v "\${fasta_prefix}-results/\${fasta_prefix}-summary.tab" ${prefix}.agrvate.tsv
    
    # These outputs are not edited further
    cut -f 2 ${prefix}.agrvate.tsv | tail -n 1 | tee AGR_GROUP
    cut -f 3 ${prefix}.agrvate.tsv | tail -n 1 | tee AGR_MATCH_SCORE
    # Parse output summary TSV -- make temporary files for noclobber rule in nf
    cut -f 4 ${prefix}.agrvate.tsv | tail -n 1 | tee AGR_CANONICAL.tmp
    cut -f 5 ${prefix}.agrvate.tsv | tail -n 1 | tee AGR_MULTIPLE.tmp
    cut -f 6 ${prefix}.agrvate.tsv | tail -n 1 | tee AGR_NUM_FRAMESHIFTS.tmp
    
    # Edit output string AGR_CANONICAL to be more informative
    # https://github.com/VishnuRaghuram94/AgrVATE#results
    if [[ \$(cat AGR_CANONICAL.tmp) == 1 ]]; then
        echo "1. canonical agrD" > AGR_CANONICAL
    elif [[ \$(cat AGR_CANONICAL.tmp) == 0 ]]; then
        echo "0. non-canonical agrD" > AGR_CANONICAL
    elif [[ \$(cat AGR_CANONICAL.tmp) == "u" ]]; then
        echo "u. unknown agrD" > AGR_CANONICAL
    else 
        echo "result unrecognized, please see summary agrvate TSV file" > AGR_CANONICAL
    fi
    
    # Edit output string AGR_MULTIPLE to be more informative
    # https://github.com/VishnuRaghuram94/AgrVATE#results
    if [[ \$(cat AGR_MULTIPLE.tmp) == "s" ]]; then
        echo "s. single agr group found" > AGR_MULTIPLE
    elif [[ \$(cat AGR_MULTIPLE.tmp) == "m" ]]; then
        echo "m. multiple agr groups found" > AGR_MULTIPLE
    elif [[ \$(cat AGR_MULTIPLE.tmp) == "u" ]]; then
        echo "u. unknown agr groups found" > AGR_MULTIPLE
    else 
        echo "result unrecognized, please see summary agrvate TSV file" > AGR_MULTIPLE
    fi
    
    # If AGR_NUM_FRAMESHIFTS is unknown, edit output string to be more informative
    # https://github.com/VishnuRaghuram94/AgrVATE#results
    if [[ \$(cat AGR_NUM_FRAMESHIFTS.tmp) == "u" ]]; then
        echo "u or unknown; agr operon not extracted" > AGR_NUM_FRAMESHIFTS
    else
        cp AGR_NUM_FRAMESHIFTS.tmp AGR_NUM_FRAMESHIFTS
    fi
    
    # Create tarball of all output files
    tar -czvf ${prefix}.agrvate.tar.gz "\${fasta_prefix}-results/"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        agrvate: \$(agrvate -v 2>&1 | sed 's/agrvate v//;')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.agrvate.tsv
    touch ${prefix}.agrvate.tar
    gzip ${prefix}.agrvate.tar
    echo "unknown" > AGR_GROUP
    echo "unknown" > AGR_MATCH_SCORE
    echo "result unrecognized, please see summary agrvate TSV file" > AGR_CANONICAL
    echo "result unrecognized, please see summary agrvate TSV file" > AGR_MULTIPLE
    echo "u or unknown; agr operon not extracted" > AGR_NUM_FRAMESHIFTS

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        agrvate: \$(agrvate -v 2>&1 | sed 's/agrvate v//;')
    END_VERSIONS
    """
}