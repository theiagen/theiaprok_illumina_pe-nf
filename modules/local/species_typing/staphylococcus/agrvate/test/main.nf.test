nextflow_process {
    name "Test Process AGRVATE"
    script "../main.nf"
    process "AGRVATE"

    test("AGRVATE - test") {
        setup {
            run("GUNZIP") {
                script "../../../../../nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'test_sample'],
                        file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/839/195/GCF_001839195.1_ASM183919v1/GCF_001839195.1_ASM183919v1_genomic.fna.gz', checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = GUNZIP.out.gunzip
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.agrvate_summary,
                    process.out.agrvate_results,
                    process.out.agrvate_agr_group,
                    process.out.agrvate_agr_match_score,
                    process.out.agrvate_agr_canonical,
                    process.out.agrvate_agr_multiple,
                    process.out.agrvate_agr_num_frameshifts,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("AGRVATE  - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/839/195/GCF_001839195.1_ASM183919v1/GCF_001839195.1_ASM183919v1_genomic.fna.gz', checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}