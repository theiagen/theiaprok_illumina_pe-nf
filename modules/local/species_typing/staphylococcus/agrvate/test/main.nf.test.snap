{
    "AGRVATE  - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.agrvate.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.agrvate.tar.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_GROUP:md5,d9f7497c382d9ee2709f9d1b560aecaf"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_MATCH_SCORE:md5,d9f7497c382d9ee2709f9d1b560aecaf"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_CANONICAL:md5,659c499e9573b6f86feee58e7165e7b0"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_MULTIPLE:md5,659c499e9573b6f86feee58e7165e7b0"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_NUM_FRAMESHIFTS:md5,b6cae007c6ca87d25b2ea96db64c7f5d"
                    ]
                ],
                "7": [
                    "versions.yml:md5,fbf8f6ba5629f99370082f5f837a7a2e"
                ],
                "agrvate_agr_canonical": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_CANONICAL:md5,659c499e9573b6f86feee58e7165e7b0"
                    ]
                ],
                "agrvate_agr_group": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_GROUP:md5,d9f7497c382d9ee2709f9d1b560aecaf"
                    ]
                ],
                "agrvate_agr_match_score": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_MATCH_SCORE:md5,d9f7497c382d9ee2709f9d1b560aecaf"
                    ]
                ],
                "agrvate_agr_multiple": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_MULTIPLE:md5,659c499e9573b6f86feee58e7165e7b0"
                    ]
                ],
                "agrvate_agr_num_frameshifts": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "AGR_NUM_FRAMESHIFTS:md5,b6cae007c6ca87d25b2ea96db64c7f5d"
                    ]
                ],
                "agrvate_results": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.agrvate.tar.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "agrvate_summary": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.agrvate.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fbf8f6ba5629f99370082f5f837a7a2e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-25T13:38:46.850827258"
    },
    "AGRVATE - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.agrvate.tsv:md5,6e87b5fc3b52f82f19c47ae8a0af41ba"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.agrvate.tar.gz:md5,427de7c66cfbd0514577dc7752dde52b"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "AGR_GROUP:md5,348a272f9cde0d186e82348e85aefab6"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "AGR_MATCH_SCORE:md5,aa6ed9e0f26a6eba784aae8267df1951"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "AGR_CANONICAL:md5,81c2592c1a13361950b19c2fd2e84189"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "AGR_MULTIPLE:md5,796329042ec830f5c9f0c8e6aae29def"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "AGR_NUM_FRAMESHIFTS:md5,897316929176464ebc9ad085f31e7284"
                ]
            ],
            [
                "versions.yml:md5,fbf8f6ba5629f99370082f5f837a7a2e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-25T13:38:34.992780557"
    }
}